homeassistant/components/esphome/config_flow.py

```diff
@@ -288,14 +288,56 @@ async def async_step_mqtt(
 
         return await self.async_step_discovery_confirm()
 
+    async def try_connect(self, entry: ConfigEntry) -> str | None:
+        """Try connecting to a device and return any errors."""
+        host = entry.data[CONF_HOST]
+        port = entry.data[CONF_PORT]
+        password = entry.data[CONF_PASSWORD]
+        psk = entry.data.get(CONF_NOISE_PSK, "")
+
+        zeroconf_instance = await zeroconf.async_get_instance(self.hass)
+
+        cli = APIClient(
+            host,
+            port,
+            password,
+            zeroconf_instance=zeroconf_instance,
+            noise_psk=psk,
+        )
+
+        try:
+            await cli.connect()
+        except APIConnectionError:
+            return "connection_error"
+        finally:
+            await cli.disconnect(force=True)
+
+        return None
+
     async def async_step_dhcp(
         self, discovery_info: dhcp.DhcpServiceInfo
     ) -> ConfigFlowResult:
         """Handle DHCP discovery."""
-        await self.async_set_unique_id(format_mac(discovery_info.macaddress))
-        self._abort_if_unique_id_configured(updates={CONF_HOST: discovery_info.ip})
-        # This should never happen since we only listen to DHCP requests
+        entry = await self.async_set_unique_id(format_mac(discovery_info.macaddress))
+
+        # entry should never be None since we only listen to DHCP requests
         # for configured devices.
+        assert entry is not None
+
+        # Check if the existing entry is still valid
+        error = await self.try_connect(entry)
```

Comment:
We should do something like https://github.com/home-assistant/core/blob/04101b044b2618989ab35753219b62d66c60440a/homeassistant/components/powerwall/config_flow.py#L104 so we don't probe the device if config entry is setup and communicating

Reply:
Will do.

Reply:
So I looked into this. There's two problems I'm running into: 

1) I can't seem to identify any state that is similar to powerwall's `async_last_update_was_successful` state in the esphome integration or aioesphomeapi. If you have a recommendation / pointer, please let me know.
2) I added a guard on the `config_entry.state` like in Powerwall, to only allow probing when the config_entry is in retry or failure state. However, the `test_connection_aborted_wrong_device` test in test_manager.py started failing due to that change. If I understand the test correctly, then this test makes sure that the config entry gets changed if it connects to the configured node, but finds a node with a different name there. I'll need to make sure that the probing logic correctly fails for this test. I had `connect` throw an `APIConnectionError` so far, but I don't think that is the expected behavior. Instead, the connection should succeed, but the DeviceInfo name should be `"different"` (vs the expected `"test"`). That said, the actual issue I'm running into is with guarding with `config_entry.state`. At least during test execution `config_entry.state` does not get changed to any of the failure states upon the wrong DeviceInfo being reported. I realized that I am not sure what the correct expectations here are? Maybe I am misunderstanding how this is supposed to work, but should the config_entry be set to a SETUP_ERROR or SETUP_RETRY state upon connecting to a wrong device? If not, then I cannot guard against `config_entry.state` if other parts of the integration rely on the DHCP update to be correctly processed here.

Reply:
I think I just found `entry_data.available`, which might solve for point 1) 
For point 2) I would still appreciate some clarification though.

Reply:
For #2 ESPHome always sets up an never enters a retry state because it never raises `ConfigEntryNotReady`. This is mostly for historical reasons since before all the retry logic was built-in. The only time is would enter setup error is if there was an unhandled exception in during setup. We should not plan for this case as it would be a bug that should be fixed

Reply:
Thanks. I'm not 100% sure if I understood what you're suggesting to do in the context of this PR. You're saying I should ignore `config_entry.state` for now and only depend on `entry_data.available`, since it would cause a bug otherwise? I can add a comment in the code that `config_entry.state` should be added to the check once ESPHome's config entry state management was modernized. Does that make sense?